eth-rpc/client/clientdispatch.nim

```diff
@@ -6,53 +6,46 @@ type
     awaiting: Table[string, Future[Response]]
     address: string
     port: Port
+    nextId: int64
   Response* = tuple[error: bool, result: JsonNode]
 
+
 proc newRpcClient*(): RpcClient =
   ## Creates a new ``RpcClient`` instance. 
   RpcClient(
     socket: newAsyncSocket(),
-    awaiting: initTable[string, Future[Response]]()
+    awaiting: initTable[string, Future[Response]](),
+    nextId: 1
   )
 
 proc call*(self: RpcClient, name: string, params: JsonNode): Future[Response] {.async.} =
   ## Remotely calls the specified RPC method.
-  # REVIEW: is there a reason why a simple counter is not used here?
-  # genOid takes CPU cycles and the output is larger
-  let id = $genOid()
-  let msg = %{"jsonrpc": %"2.0", "method": %name, "params": params, "id": %id}
-  # REVIEW: it would be more efficient if you append the terminating new line to
-  # the `msg` variable in-place. This way, a copy won't be performed most of the
-  # time because the string is likely to have 2 bytes of unused capacity.
-  await self.socket.send($msg & "\c\l")
+  let id = $self.nextId
+  self.nextId.inc
+  let msg = $ %{"jsonrpc": %"2.0", "method": %name, "params": params, "id": %id} & "\c\l"
+  await self.socket.send(msg)
 
-  # Completed by processMessage.
+  # completed by processMessage.
   var newFut = newFuture[Response]()
-  self.awaiting[id] = newFut  # add to awaiting responses
+  # add to awaiting responses
+  self.awaiting[id] = newFut
   result = await newFut
 
-proc isNull(node: JsonNode): bool = node.kind == JNull
-
 proc processMessage(self: RpcClient, line: string) =
   let node = parseJson(line)
   
-  # REVIEW: These shouldn't be just asserts. You cannot count
-  # that the other side implements the protocol correctly, so
-  # you must perform validation even in release builds.
-  assert node.hasKey("jsonrpc")
-  assert node["jsonrpc"].str == "2.0"
-  assert node.hasKey("id")
-  assert self.awaiting.hasKey(node["id"].str)
+  # TODO: Use more appropriate exception objects
+  if not node.hasKey("jsonrpc"): raise newException(ValueError, "Message is missing rpc version field")
```

Comment:
Minor optimisation suggestion: `hasKey("key")` and `node["key"]` both do lookups. Number of lookups can be reduced to 1 with `node{"key"}`.